tests/integration/test_ask_tell_optimization.py

```diff
@@ -0,0 +1,142 @@
+    cast(
+        List[
+            Tuple[
+                int,
+                bool,
+                Union[
+                    AcquisitionRule[TensorType, Box],
+                    AcquisitionRule[State[TensorType, TrustRegion.State], Box],
+                ],
+            ]
+        ],
+        [
+            (20, False, EfficientGlobalOptimization()),
+            (20, True, EfficientGlobalOptimization()),
+            (
+                10,
+                False,
+                EfficientGlobalOptimization(
+                    LocalPenalizationAcquisitionFunction(
+                        BRANIN_SEARCH_SPACE,
+                    ).using(OBJECTIVE),
+                    num_query_points=3,
+                ),
+            ),
+            (
+                10,
+                True,
+                EfficientGlobalOptimization(
+                    LocalPenalizationAcquisitionFunction(
+                        BRANIN_SEARCH_SPACE,
+                    ).using(OBJECTIVE),
+                    num_query_points=3,
+                ),
+            ),
+            (15, False, TrustRegion()),
+            (15, True, TrustRegion()),
+        ],
+    ),
+)
+def test_ask_tell_optimization_finds_minima_of_the_scaled_branin_function(
+    num_steps: int,
+    reload_state: bool,
+    acquisition_rule: AcquisitionRule[TensorType, SearchSpace]
+    | AcquisitionRule[State[TensorType, TrustRegion.State], Box],
+) -> None:
+    search_space = BRANIN_SEARCH_SPACE
+
+    def build_model(data: Dataset) -> GaussianProcessRegression:
+        variance = tf.math.reduce_variance(data.observations)
+        kernel = gpflow.kernels.Matern52(variance, tf.constant([0.2, 0.2], tf.float64))
+        scale = tf.constant(1.0, dtype=tf.float64)
+        kernel.variance.prior = tfp.distributions.LogNormal(
+            tf.constant(-2.0, dtype=tf.float64), scale
+        )
+        kernel.lengthscales.prior = tfp.distributions.LogNormal(
+            tf.math.log(kernel.lengthscales), scale
+        )
+        gpr = gpflow.models.GPR((data.query_points, data.observations), kernel, noise_variance=1e-5)
+        gpflow.utilities.set_trainable(gpr.likelihood, False)
+        return GaussianProcessRegression(gpr)
+
+    initial_query_points = search_space.sample(5)
+    observer = mk_observer(scaled_branin)
+    initial_data = observer(initial_query_points)
+    model = build_model(initial_data)
+
+    ask_tell = AskTellOptimizer(search_space, initial_data, model, True, acquisition_rule)
+
+    for _ in range(num_steps):
+        # two scenarios can be tested here, depending on `reload_state` parameter
+        # in first the same optimizer object is always used
+        # in second new optimizer is created at each step (e.g. from serialized state)
+        new_point = ask_tell.ask()
+
+        if reload_state:
+            state = ask_tell.get_state()
+
+        new_data = observer(new_point)
+
+        if reload_state:
```

Comment:
Not part of this PR, but would be nice to test actually serialising records to disk.

Reply:
i understand not all types of models can be serialized at the moment, so it may be a bit tricky